### ISSUE TITLE
"Pack" memory allocation

### ISSUE THREAD

**Issue by lemire** [simdjson/simdjson#94]
We do several memory allocations at once. They could be combined in a single, larger memory allocation, for possibly better performance.

**Comment by TkTech** [comment]
After this change, would it be reasonable to also lookup the free/malloc from a pointer? Python, JVM, and pretty much everything else with a runtime implement their own userspace free/malloc to more efficiently re-use memory without a context switch. Also allows bookkeeper for your allocations from the higher-level tools, and the use of debugging allocators when looking for faults. It would be nice to be able to use [PyMem_Malloc](https://docs.python.org/3/c-api/memory.html#c.PyMem_Malloc)!

**Comment by lemire** [comment]
@TkTech I expect you should already be able to hack the memory allocation and provide your own. You can always provide your own `malloc` and stuff. I bet you can already use PyMem_Malloc without modifying any code in simdjson proper. Of course, I might be wrong, but if I am, I'd like to know how.

**Comment by TkTech** [comment]
You can, but it is not portable or trivial since it typical requires symbol management and can affect other libraries you've linked to. The approach used by practically all libraries is to store pointers to them in whatever management structure they're using. For example for zlib you have `z_stream.zalloc` and `z_stream.zfree` which just point to malloc/free by default.

**Comment by lemire** [comment]
We can do that, sure.

**Comment by lemire** [comment]
Memory allocation can be a bottleneck, as exemplified by PR https://github.com/lemire/simdjson/pull/335

```
$ ./kostyacompetition jsonexamples/kostyalarge.json
simdjson                                	:    4.104 cycles per input byte (best)    4.125 cycles per input byte (avg)    0.830 GB/s (error margin: 0.004 GB/s)
simdjson (just parse)                   	:    3.770 cycles per input byte (best)    3.788 cycles per input byte (avg)    0.904 GB/s (error margin: 0.004 GB/s)
simdjson (just dom)                     	:    0.329 cycles per input byte (best)    0.332 cycles per input byte (avg)   10.347 GB/s (error margin: 0.071 GB/s)
simdjson (just parse, no alloc)         	:    2.377 cycles per input byte (best)    2.690 cycles per input byte (avg)    1.434 GB/s (error margin: 0.167 GB/s)
```


The only difference between (just parse) and  (just parse, no alloc) is the memory allocation for the parsed json. It is costing us dearly.

**Comment by lemire** [comment]
I want to recopy @jkeiser comment...

"Right now, we do [3 allocations per parser](https://github.com/simdjson/simdjson/blob/master/include/simdjson/inline/document.h#L223-L224) and [2 allocations per document](https://github.com/simdjson/simdjson/blob/master/include/simdjson/inline/document.h#L223-L224), for a total of 5. Instead, we could do 1 megaallocation for the parser, and another for the document; and the pointers would each just take their portion of that allocated memory.

This would probably allow us to reference a document with a single pointer-sized object, which would help simdjson bind to languages that can store a single pointer efficiently (but can't store 2 efficiently)."

**Comment by KubaO** [comment]
> @TkTech I expect you should already be able to hack the memory allocation and provide your own. You can always provide your own `malloc` and stuff. I bet you can already use PyMem_Malloc without modifying any code in simdjson proper. Of course, I might be wrong, but if I am, I'd like to know how.

In my experience, substituting the symbol named literally `malloc` and `free` is a salty, tear-filled minefield. Either it's explicitly supported using library-specific weak functions/symbols that forward to `malloc` and `free`, or it's not supported. It's all fine and dandy when you need to substitute these symbols in one library, but what about having 10 libraries, where you only want to substitute half? It's a lost cause unless said libraries use their own dedicated symbols that the user can override.

**Comment by lemire** [comment]
@KubaO  That's a fair point, and I have created an issue for that purpose: https://github.com/simdjson/simdjson/issues/1017

**Comment by jkeiser** [comment]
I can't find my results, but I did try doing a single allocation, and got a very noticeable drop in performance, on the order of 20%, so I abandoned my attempt. My theory was that since each of our allocations is larger than  it needs to be, having separate allocations let's the system allocate less memory overall, saving time--if we use 20 bytes out of the first million, but then use another 20 bytes,.the system probably has to fill in the million bytes in between. It could be I was doing it wrong, though.
